source/routing/journey_pattern_container.cpp

```diff
@@ -0,0 +1,176 @@
+         it1 !=  vj1.stop_time_list.end() && it2 !=  vj2.stop_time_list.end();
+         ++it1, ++it2) {
+        if (it1->arrival_time != it2->arrival_time || it1->departure_time != it2->departure_time) {
+            return false;
+        }
+    }
+    return true;
+}
+
+template<typename VJ> static bool
+overtake(const VJ& vj, const std::vector<const VJ*> vjs) {
+    if (vj.stop_time_list.empty()) { return false; }
+    for (const VJ* cur_vj: vjs) {
+        assert(vj.stop_time_list.size() == cur_vj->stop_time_list.size());
+
+        // if the validity patterns do not overlap, it can't overtake
+        const auto levels = {nt::RTLevel::Theoric, nt::RTLevel::Adapted, nt::RTLevel::RealTime};
+        const bool dont_overlap = boost::algorithm::all_of(levels, [&](nt::RTLevel lvl) {
+                if (vj.validity_patterns[lvl] == nullptr) { return true; }
+                if (cur_vj->validity_patterns[lvl] == nullptr) { return true; }
+                return (vj.validity_patterns[lvl]->days & cur_vj->validity_patterns[lvl]->days).none();
+            });
+        if (dont_overlap) { continue; }
+
+        // if the stop times are the same, they don't overtake
+        if (st_are_equals(vj, *cur_vj)) { continue; }
+
+        const bool vj_is_first =
+            vj.stop_time_list.front().departure_time < cur_vj->stop_time_list.front().departure_time;
+        const VJ& vj1 = vj_is_first ? vj : *cur_vj;
+        const VJ& vj2 = vj_is_first ? *cur_vj : vj;
+        for (auto it1 = vj1.stop_time_list.begin(), it2 = vj2.stop_time_list.begin();
+             it1 !=  vj1.stop_time_list.end() && it2 !=  vj2.stop_time_list.end();
+             ++it1, ++it2) {
+            if (it1->arrival_time >= it2->arrival_time ||
+                it1->departure_time >= it2->departure_time) {
+                return true;
+            }            
+        }
+    }
+    return false;
+}
+
+template<typename VJ>
+void JourneyPatternContainer::add_vj(const VJ& vj) {
+    // Get the existing jps according to the key.
+    const auto key = make_key(vj);
+    auto& jps = map[key];
+
+    // The vj must not overtake the other vj in its jp, thus searching
+    // for one satisfying jp.
+    for (auto& jp_idx: jps) {
+        auto& jp = get_mut(jp_idx);
+        auto& vjs = jp.template get_vjs<VJ>();
+        if (! overtake(vj, vjs)) {
+            // Great, found a valid jp, inserting and stopping
+            vjs.push_back(&vj);
+            jp_from_vj[VjIdx(vj)] = jp_idx;
+            return;
+        }
+    }
+
+    // We did not find a jp, creating a new one.
+    const auto jp_idx = make_jp(key);
+    jps.push_back(jp_idx);
+    jps_from_route[key.route_idx].push_back(jp_idx);
+    get_mut(jp_idx).template get_vjs<VJ>().push_back(&vj);
+    jp_from_vj[VjIdx(vj)] = jp_idx;
+}
+
+JpIdx JourneyPatternContainer::make_jp(const JpKey& key) {
+    const auto jp_idx = JpIdx(jps.size());
+    jps.emplace_back();
```

Comment:
I'm a little bothered by doing so :)
couldn't you construct a jp and move it into jps?